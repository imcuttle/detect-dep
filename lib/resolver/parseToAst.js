/**
 * @file parseToAst
 * @author Cuttle Cong
 * @date 2018/2/20
 * @description
 */

const babylon = require('@babel/parser')

const opts = {
  allowImportExportEverywhere: true,
  allowAwaitOutsideFunction: true,
  sourceType: 'module',
  locations: false,
  plugins: [
    'jsx',
    // 'flow',
    'typescript',

    'asyncGenerators',
    'bigInt',
    'classProperties',
    'classPrivateProperties',
    'classPrivateMethods',
    'decorators-legacy',
    'doExpressions',
    'dynamicImport',
    'exportDefaultFrom',
    'exportNamespaceFrom',
    'functionBind',
    'functionSent',
    'importMeta',
    'logicalAssignment',
    'nullishCoalescingOperator',
    'numericSeparator',
    'objectRestSpread',
    'optionalCatchBinding',
    'optionalChaining',
    'partialApplication',
    // 'pipelineOperator',
    'throwExpressions',
    'topLevelAwait'
    // 'asyncGenerators',
    // 'classConstructorCall',
    // 'classProperties',
    // 'decorators',
    // 'doExpressions',
    // 'exportExtensions',
    // 'flow',
    // 'functionBind',
    // 'functionSent',
    // 'jsx',
    // 'objectRestSpread',
    // 'dynamicImport'
  ]
}

module.exports = function parseToAst(source) {
  return babylon.parse(source, opts)
}
