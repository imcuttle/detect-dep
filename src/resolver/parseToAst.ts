/**
 * @file parseToAst
 * @author Cuttle Cong
 * @date 2018/2/20
 * @description
 */

import { ParserOptions, parse } from '@babel/parser'

const defaultOpts: ParserOptions = {
  allowImportExportEverywhere: true,
  allowAwaitOutsideFunction: true,
  sourceType: 'module',
  // locations: false,
  plugins: [
    'jsx',
    // 'flow',
    'typescript',

    'asyncGenerators',
    'bigInt',
    'classProperties',
    'classPrivateProperties',
    'classPrivateMethods',
    'decorators-legacy',
    'doExpressions',
    'dynamicImport',
    'exportDefaultFrom',
    'exportNamespaceFrom',
    'functionBind',
    'functionSent',
    'importMeta',
    'logicalAssignment',
    'nullishCoalescingOperator',
    'numericSeparator',
    'objectRestSpread',
    'optionalCatchBinding',
    'optionalChaining',
    'partialApplication',
    // 'pipelineOperator',
    'throwExpressions',
    'topLevelAwait'
  ]
}

export default function parseToAst(source, opts?: ParserOptions) {
  return parse(source, {
    ...defaultOpts,
    ...opts
  })
}
