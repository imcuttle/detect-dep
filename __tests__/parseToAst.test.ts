/**
 * @file parseToAst
 * @author Cuttle Cong
 * @date 2018/2/20
 * @description
 */
import parseToAst from '../src/resolver/parseToAst'

function testSnapshot(codes) {
  expect(parseToAst(codes.join('\n'))).toMatchSnapshot()
}

describe('parseToAst', function () {
  it('JSX', function () {
    testSnapshot(["import React from 'react'", 'const a = <div></div>'])
  })

  it('classProperties', function () {
    testSnapshot(['class View {', '  a = 2', '  ab = () => {}', '  static v = () => {}', '}'])
  })
})
