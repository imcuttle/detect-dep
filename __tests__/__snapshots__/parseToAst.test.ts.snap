// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseToAst JSX 1`] = `
Node {
  "comments": Array [],
  "end": 47,
  "errors": Array [],
  "extra": undefined,
  "innerComments": undefined,
  "leadingComments": undefined,
  "loc": SourceLocation {
    "end": Position {
      "column": 21,
      "line": 2,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "end": 25,
        "extra": undefined,
        "importKind": "value",
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 25,
            "line": 1,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "range": undefined,
        "source": Node {
          "end": 25,
          "extra": Object {
            "raw": "'react'",
            "rawValue": "react",
          },
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 25,
              "line": 1,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 18,
              "line": 1,
            },
          },
          "range": undefined,
          "start": 18,
          "trailingComments": undefined,
          "type": "StringLiteral",
          "value": "react",
        },
        "specifiers": Array [
          Node {
            "end": 12,
            "extra": undefined,
            "innerComments": undefined,
            "leadingComments": undefined,
            "loc": SourceLocation {
              "end": Position {
                "column": 12,
                "line": 1,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 7,
                "line": 1,
              },
            },
            "local": Node {
              "end": 12,
              "extra": undefined,
              "innerComments": undefined,
              "leadingComments": undefined,
              "loc": SourceLocation {
                "end": Position {
                  "column": 12,
                  "line": 1,
                },
                "filename": undefined,
                "identifierName": "React",
                "start": Position {
                  "column": 7,
                  "line": 1,
                },
              },
              "name": "React",
              "range": undefined,
              "start": 7,
              "trailingComments": undefined,
              "type": "Identifier",
            },
            "range": undefined,
            "start": 7,
            "trailingComments": undefined,
            "type": "ImportDefaultSpecifier",
          },
        ],
        "start": 0,
        "trailingComments": undefined,
        "type": "ImportDeclaration",
      },
      Node {
        "declarations": Array [
          Node {
            "end": 47,
            "extra": undefined,
            "id": Node {
              "end": 33,
              "extra": undefined,
              "innerComments": undefined,
              "leadingComments": undefined,
              "loc": SourceLocation {
                "end": Position {
                  "column": 7,
                  "line": 2,
                },
                "filename": undefined,
                "identifierName": "a",
                "start": Position {
                  "column": 6,
                  "line": 2,
                },
              },
              "name": "a",
              "range": undefined,
              "start": 32,
              "trailingComments": undefined,
              "type": "Identifier",
            },
            "init": Node {
              "children": Array [],
              "closingElement": Node {
                "end": 47,
                "extra": undefined,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 21,
                    "line": 2,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 15,
                    "line": 2,
                  },
                },
                "name": Node {
                  "end": 46,
                  "extra": undefined,
                  "innerComments": undefined,
                  "leadingComments": undefined,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 20,
                      "line": 2,
                    },
                    "filename": undefined,
                    "identifierName": undefined,
                    "start": Position {
                      "column": 17,
                      "line": 2,
                    },
                  },
                  "name": "div",
                  "range": undefined,
                  "start": 43,
                  "trailingComments": undefined,
                  "type": "JSXIdentifier",
                },
                "range": undefined,
                "start": 41,
                "trailingComments": undefined,
                "type": "JSXClosingElement",
              },
              "end": 47,
              "extra": undefined,
              "innerComments": undefined,
              "leadingComments": undefined,
              "loc": SourceLocation {
                "end": Position {
                  "column": 21,
                  "line": 2,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 10,
                  "line": 2,
                },
              },
              "openingElement": Node {
                "attributes": Array [],
                "end": 41,
                "extra": undefined,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 2,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 10,
                    "line": 2,
                  },
                },
                "name": Node {
                  "end": 40,
                  "extra": undefined,
                  "innerComments": undefined,
                  "leadingComments": undefined,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 14,
                      "line": 2,
                    },
                    "filename": undefined,
                    "identifierName": undefined,
                    "start": Position {
                      "column": 11,
                      "line": 2,
                    },
                  },
                  "name": "div",
                  "range": undefined,
                  "start": 37,
                  "trailingComments": undefined,
                  "type": "JSXIdentifier",
                },
                "range": undefined,
                "selfClosing": false,
                "start": 36,
                "trailingComments": undefined,
                "type": "JSXOpeningElement",
              },
              "range": undefined,
              "start": 36,
              "trailingComments": undefined,
              "type": "JSXElement",
            },
            "innerComments": undefined,
            "leadingComments": undefined,
            "loc": SourceLocation {
              "end": Position {
                "column": 21,
                "line": 2,
              },
              "filename": undefined,
              "identifierName": undefined,
              "start": Position {
                "column": 6,
                "line": 2,
              },
            },
            "range": undefined,
            "start": 32,
            "trailingComments": undefined,
            "type": "VariableDeclarator",
          },
        ],
        "end": 47,
        "extra": undefined,
        "innerComments": undefined,
        "kind": "const",
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 21,
            "line": 2,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "range": undefined,
        "start": 26,
        "trailingComments": undefined,
        "type": "VariableDeclaration",
      },
    ],
    "directives": Array [],
    "end": 47,
    "extra": undefined,
    "innerComments": undefined,
    "interpreter": null,
    "leadingComments": undefined,
    "loc": SourceLocation {
      "end": Position {
        "column": 21,
        "line": 2,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "range": undefined,
    "sourceType": "module",
    "start": 0,
    "trailingComments": undefined,
    "type": "Program",
  },
  "range": undefined,
  "start": 0,
  "trailingComments": undefined,
  "type": "File",
}
`;

exports[`parseToAst classProperties 1`] = `
Node {
  "comments": Array [],
  "end": 60,
  "errors": Array [],
  "extra": undefined,
  "innerComments": undefined,
  "leadingComments": undefined,
  "loc": SourceLocation {
    "end": Position {
      "column": 1,
      "line": 5,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "body": Node {
          "body": Array [
            Node {
              "computed": false,
              "end": 20,
              "extra": undefined,
              "innerComments": undefined,
              "key": Node {
                "end": 16,
                "extra": undefined,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 3,
                    "line": 2,
                  },
                  "filename": undefined,
                  "identifierName": "a",
                  "start": Position {
                    "column": 2,
                    "line": 2,
                  },
                },
                "name": "a",
                "range": undefined,
                "start": 15,
                "trailingComments": undefined,
                "type": "Identifier",
              },
              "leadingComments": undefined,
              "loc": SourceLocation {
                "end": Position {
                  "column": 7,
                  "line": 2,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 2,
                  "line": 2,
                },
              },
              "range": undefined,
              "start": 15,
              "static": false,
              "trailingComments": undefined,
              "type": "ClassProperty",
              "value": Node {
                "end": 20,
                "extra": Object {
                  "raw": "2",
                  "rawValue": 2,
                },
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 7,
                    "line": 2,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 6,
                    "line": 2,
                  },
                },
                "range": undefined,
                "start": 19,
                "trailingComments": undefined,
                "type": "NumericLiteral",
                "value": 2,
              },
            },
            Node {
              "computed": false,
              "end": 36,
              "extra": undefined,
              "innerComments": undefined,
              "key": Node {
                "end": 25,
                "extra": undefined,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 4,
                    "line": 3,
                  },
                  "filename": undefined,
                  "identifierName": "ab",
                  "start": Position {
                    "column": 2,
                    "line": 3,
                  },
                },
                "name": "ab",
                "range": undefined,
                "start": 23,
                "trailingComments": undefined,
                "type": "Identifier",
              },
              "leadingComments": undefined,
              "loc": SourceLocation {
                "end": Position {
                  "column": 15,
                  "line": 3,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 2,
                  "line": 3,
                },
              },
              "range": undefined,
              "start": 23,
              "static": false,
              "trailingComments": undefined,
              "type": "ClassProperty",
              "value": Node {
                "async": false,
                "body": Node {
                  "body": Array [],
                  "directives": Array [],
                  "end": 36,
                  "extra": undefined,
                  "innerComments": undefined,
                  "leadingComments": undefined,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 15,
                      "line": 3,
                    },
                    "filename": undefined,
                    "identifierName": undefined,
                    "start": Position {
                      "column": 13,
                      "line": 3,
                    },
                  },
                  "range": undefined,
                  "start": 34,
                  "trailingComments": undefined,
                  "type": "BlockStatement",
                },
                "end": 36,
                "extra": undefined,
                "generator": false,
                "id": null,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 3,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 7,
                    "line": 3,
                  },
                },
                "params": Array [],
                "range": undefined,
                "start": 28,
                "trailingComments": undefined,
                "type": "ArrowFunctionExpression",
              },
            },
            Node {
              "computed": false,
              "end": 58,
              "extra": undefined,
              "innerComments": undefined,
              "key": Node {
                "end": 47,
                "extra": undefined,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 10,
                    "line": 4,
                  },
                  "filename": undefined,
                  "identifierName": "v",
                  "start": Position {
                    "column": 9,
                    "line": 4,
                  },
                },
                "name": "v",
                "range": undefined,
                "start": 46,
                "trailingComments": undefined,
                "type": "Identifier",
              },
              "leadingComments": undefined,
              "loc": SourceLocation {
                "end": Position {
                  "column": 21,
                  "line": 4,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 2,
                  "line": 4,
                },
              },
              "range": undefined,
              "start": 39,
              "static": true,
              "trailingComments": undefined,
              "type": "ClassProperty",
              "value": Node {
                "async": false,
                "body": Node {
                  "body": Array [],
                  "directives": Array [],
                  "end": 58,
                  "extra": undefined,
                  "innerComments": undefined,
                  "leadingComments": undefined,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 21,
                      "line": 4,
                    },
                    "filename": undefined,
                    "identifierName": undefined,
                    "start": Position {
                      "column": 19,
                      "line": 4,
                    },
                  },
                  "range": undefined,
                  "start": 56,
                  "trailingComments": undefined,
                  "type": "BlockStatement",
                },
                "end": 58,
                "extra": undefined,
                "generator": false,
                "id": null,
                "innerComments": undefined,
                "leadingComments": undefined,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 21,
                    "line": 4,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 13,
                    "line": 4,
                  },
                },
                "params": Array [],
                "range": undefined,
                "start": 50,
                "trailingComments": undefined,
                "type": "ArrowFunctionExpression",
              },
            },
          ],
          "end": 60,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 5,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 11,
              "line": 1,
            },
          },
          "range": undefined,
          "start": 11,
          "trailingComments": undefined,
          "type": "ClassBody",
        },
        "end": 60,
        "extra": undefined,
        "id": Node {
          "end": 10,
          "extra": undefined,
          "innerComments": undefined,
          "leadingComments": undefined,
          "loc": SourceLocation {
            "end": Position {
              "column": 10,
              "line": 1,
            },
            "filename": undefined,
            "identifierName": "View",
            "start": Position {
              "column": 6,
              "line": 1,
            },
          },
          "name": "View",
          "range": undefined,
          "start": 6,
          "trailingComments": undefined,
          "type": "Identifier",
        },
        "innerComments": undefined,
        "leadingComments": undefined,
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 5,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 1,
          },
        },
        "range": undefined,
        "start": 0,
        "superClass": null,
        "trailingComments": undefined,
        "type": "ClassDeclaration",
      },
    ],
    "directives": Array [],
    "end": 60,
    "extra": undefined,
    "innerComments": undefined,
    "interpreter": null,
    "leadingComments": undefined,
    "loc": SourceLocation {
      "end": Position {
        "column": 1,
        "line": 5,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "range": undefined,
    "sourceType": "module",
    "start": 0,
    "trailingComments": undefined,
    "type": "Program",
  },
  "range": undefined,
  "start": 0,
  "trailingComments": undefined,
  "type": "File",
}
`;
